src/os/bluestore/BitmapFreelistManager.cc

```diff
@@ -0,0 +1,399 @@
+    txn->set(prefix, k, bl);
+  }
+  _init_misc();
+  return 0;
+}
+
+int BitmapFreelistManager::init()
+{
+  dout(1) << __func__ << dendl;
+
+  KeyValueDB::Iterator it = kvdb->get_iterator(prefix);
+  it->lower_bound(string());
+
+  // load meta
+  while (it->valid()) {
+    if (it->key()[0] != '_')
+      break;
+    string k = it->key().substr(1);
+    if (k == "bytes_per_block") {
+      bufferlist bl = it->value();
+      bufferlist::iterator p = bl.begin();
+      ::decode(bytes_per_block, p);
+      dout(10) << __func__ << " bytes_per_block " << bytes_per_block << dendl;
+    } else if (k == "blocks") {
+      bufferlist bl = it->value();
+      bufferlist::iterator p = bl.begin();
+      ::decode(blocks, p);
+      dout(10) << __func__ << " blocks " << blocks << dendl;
+    } else if (k == "blocks_per_key") {
+      bufferlist bl = it->value();
+      bufferlist::iterator p = bl.begin();
+      ::decode(blocks_per_key, p);
+      dout(10) << __func__ << " blocks_per_key " << blocks_per_key << dendl;
+    } else {
+      derr << __func__ << " unrecognized meta " << k << dendl;
+      return -EIO;
+    }
+    it->next();
+  }
+
+  dout(10) << __func__
+	   << " bytes_per_block " << bytes_per_block
+	   << " blocks " << blocks
+	   << " blocks_per_key " << blocks_per_key
+	   << dendl;
+  _init_misc();
+  return 0;
+}
+
+void BitmapFreelistManager::_init_misc()
+{
+  bufferptr z(blocks_per_key >> 3);
+  memset(z.c_str(), 0xff, z.length());
+  all_set_bl.clear();
+  all_set_bl.append(z);
+
+  block_mask = ~(bytes_per_block - 1);
+
+  uint64_t bytes_per_key = bytes_per_block * blocks_per_key;
+  key_mask = ~(bytes_per_key - 1);
+  dout(10) << __func__ << " bytes_per_key " << bytes_per_key
+	   << ", key_mask " << key_mask
+	   << dendl;
+}
+
+void BitmapFreelistManager::shutdown()
+{
+  dout(1) << __func__ << dendl;
+}
+
+void BitmapFreelistManager::enumerate_reset()
+{
+  std::lock_guard<std::mutex> l(lock);
+  enumerate_offset = 0;
+  enumerate_bl.clear();
+}
+
+int get_next_clear_bit(bufferlist& bl, int start)
+{
+  const char *p = bl.c_str();
+  int bits = bl.length() << 3;
+  while (start < bits) {
+    int byte = start >> 3;
+    unsigned char mask = 1 << (7 - (start & 7));
+    if ((p[byte] & mask) == 0) {
+      return start;
+    }
+    ++start;
+  }
+  return -1; // not found
+}
+
+int get_next_set_bit(bufferlist& bl, int start)
+{
+  const char *p = bl.c_str();
+  int bits = bl.length() << 3;
+  while (start < bits) {
+    int byte = start >> 3;
+    unsigned char mask = 1 << (7 - (start & 7));
+    if (p[byte] & mask) {
+      return start;
+    }
+    ++start;
+  }
+  return -1; // not found
+}
+
+bool BitmapFreelistManager::enumerate_next(uint64_t *offset, uint64_t *length)
+{
+  std::lock_guard<std::mutex> l(lock);
+
+  // initial base case is a bit awkward
+  if (enumerate_offset == 0) {
+    dout(10) << __func__ << " start" << dendl;
+    enumerate_p = kvdb->get_iterator(prefix);
+    enumerate_p->lower_bound(string("o"));
```

Comment:
why shall we removing this...  It means we care about the offset key(begins with "o") only... Otherwise we will locate at the meta key which begins with a  "_" here, which is wrong.